test/evaluate/folding03.f90

```diff
@@ -0,0 +1,305 @@
+! Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+! Test operation folding edge case (both expected value and messages)
+! These tests make assumptions regarding real(4) and integer(4) extrema.
+
+#define TEST_ISNAN(v) logical, parameter :: test_##v =.NOT.(v.EQ.v)
+
+! TODO: 2 behaviors to clarify:
+
+!  Biggest negative literal integer(4): -(2**31) or -(2**31-1) ?
+!   -> Fortran standard says representable value set must be in [-HUGE, HUGE].
+!      For f18 integer(4) that means [-(2**31-1), 2**31-1].
+!   -> Current f18 behavior is to emit an error for -2147483648_4 literal (-2**31), but not for
+!      -2147483647_4 -1_4 which is folded to -2147483648_4 without errors/warnings.
+!  It seems to me that biggest negative literal and representable value should
+!  be the same.
+!   -> pgfortran and ifort are OK with -2147483648_4 literal.
+!   -> gfortran gives an error for both -2147483648_4 unless -fno-range-check is used.
+!      It never raises errors for -2147483647_4 -1_4 and fold it to -2147483648_4.
+
+! Proposal: add option to control literal range. Make -2147483648_4 accepted by
+! default?
+
+! How to handle 0_4/0_4:
+!  -> raise an error (pgfortran, gfortran, ifort)
+!  -> warning and return 2147483647_4 (currently f18)
+!  -> warning and return 0_4 (maybe power HW (not tested))
+! Actually, gfortran, pgfortran and ifort forbid integer divide by zero in all
+! case.
```

Comment:
Division by zero in the computation of the value of an integer `PARAMETER` or other constant expression should be fatal, I think.  So should overflow (including `MAX_NEGATIVE_INTEGER/(-1)`) in division and other operations.  But these exceptional cases shouldn't necessarily be fatal in folded expressions outside constant expressions, just warnings.

Reply:
Right, then the folding errors have to be handled a bit differently than they currently are. The folding library does not know if the expression being folded is for a constant or not. So, it should report division by zero and overflows without judging of the severity and let semantics make it fatal or not, alternatively, the folding context could be configured before folding to indicate if these cases should be fatal or not. I would go for the second option to avoid having to add code handling errors every time Fold is called.
Should I do this now or should we do this when improving the error/warning system to classify warnings and give the option to make them fatal ?

Reply:
Folding should emit a warning and then return the original expression (possibly reduced to something canonical like 0/0).  In contexts where a constant expression is required, the fact that the result is not a constant will elicit the fatal error.

Reply:
> Folding should emit a warning and then return the original expression (possibly reduced to something canonical like 0/0)

OK, that sounds even easier. Here is the list of operations that I will then modify in fold.cc to return the original expression:
- Integer division
   - division by zero
   - overflow
-  Integer addition, subtraction, multiplication, power, negation
    - overflows

Real/Complex operations stay unchanged (NaN and +/-inf are here to propagate folding issues).

Do you agree with the list?

Reply:
I don't think that we should return the original expression in case of integer overflow -- just return the truncated value.  That matches what would have taken place at runtime.

Reply:
OK, new commits make integer division by zero in constants a fatal error.
I first went the wrong path of expecting `Constant<T` in such expression.... I should have read `IsConstantExpr` comment more carefully... I took the opportunity to remove what looked like to me an useless refold in  `EvaluateConvertedExpr` (resolve-names.cc).